### PR TITLE
kobo: Added Libra 2 and Elipsa and Sage

### PR DESCRIPTION
**Pull request by jucor**
Pull request to solve #5.
@NiLuJe, as you kindly did for the Elipsa in #4 , could you please check the coversize thing? I am an absolute noob at Kobo tweaking, not having the clues to check the firmware. Getting involved because I know @pgaskin is completely slammed at the moment, so I would very much appreciate help from someone who knows what they're doing :) 
Thank you!

### CODE REVIEW

**Comment by NiLuJe** [comment]
Might be worth adding both the Elipsa & the Sage while you're in there ;).

c.f., https://github.com/kovidgoyal/calibre/blob/7e645a0e30540261afc6f337291a3f07675323b4/src/calibre/devices/kobo/driver.py#L1461-L1508 & https://github.com/kovidgoyal/calibre/blob/7e645a0e30540261afc6f337291a3f07675323b4/src/calibre/devices/kobo/driver.py#L3463-L3506 for the cover sizes ;).

**Comment by jucor** [comment]
Thanks @NiLuJe ! Happy to add both indeed while I'm at it  -- will it be a problem if that's redundant with #4 ?

**Comment by jucor** [comment]
So, thanks to the links you sent @NiLuJe , it looks like I've got the cover sizes right for the Libra 2, i.e. the same as for the Libra H2O. Pfew :) Now looking at Elipsa and Sage.

**Comment by jucor** [comment]
Alright, Sage and Elipsa added too. Thanks @NiLuJe for the resolutions, super helpful! Now just need to find back my GPG keys, renew them, and sign my commits...

**Comment by pgaskin** [comment]
> Now just need to find back my GPG keys, renew them, and sign my commits...

Don't bother with that now; I've disabled the branch protection restriction for signed commits. It didn't require having all PR commits signed (as opposed to just the final merge/squash) back when I enabled it.

**Comment by marcus-crane** [comment]
I'd be interested to know if there's any material on how to confirm cover sizes? I briefly had a look at doing a similar PR myself but I didn't have any luck getting a slightly modified Kobo firmware to work (enabling some debug output from memory)

**Comment by jucor** [comment]
@marcus-crane Thanks for raising the question. Well, for the cover sizes, I trust  Calibre's code kindly provided by @NiLuJe  in https://github.com/pgaskin/koboutils/pull/7#issuecomment-1059815564 . This is whom @pgaskin himself asked to check his own work in #4. And with good reason: the Calibre source allowed to fix the resolution originally guesssed for the Elipsa. I have also checked mobilereads and its wiki for code names and hardware versions. There's a strong community there -- see for example [this list of codenames](https://www.mobileread.com/forums/showthread.php?t=325471) and [this list of hardware versions](https://wiki.mobileread.com/wiki/Kobo_Firmware_Releases#Hardware_versions).

I clearly admit that I do not know how to dig into the Kobo firmware myself, not without spending hours learning: my reverse-engineering days in Assembly are from the previous century, around 1997 :)  
I'll also admit that I have not compiled nor tested the above on my Libra 2, for lack of being familiar with Go toolchain. Granted, this is probably one of the least-tested PRs I've ever done -- luckily it's also one of the clearest (thus simplest) codes I've worked on :)

In the worst-case scenario, say we have a the resolution wrong for some of these new models. It's still  better than *not* supporting those models, and will allow any error to surface and make its way to the Github issues eventually.

Knowing that @pgaskin is currently very busy,  I'd be in favour of releasing this in spite of the lack of deep testing -- better a possibly-imperfect solution than no solution. Or am I being too careless?

**Comment by jucor** [comment]
@pgaskin If the above caveats didn't put you off, and if (or when) you have a minute in the current busy times, I'd be super grateful if you would merge and make a release with this. I can then get KoboUNCaGed to use it :) 
But of course, when you have the time -- good luck with what is keeping you so busy!

**Comment by marcus-crane** [comment]
I'm also looking forward to Libra 2 support being added (I have one myself). I use koboutils in https://github.com/marcus-crane/october where I patched in placeholder support for devices not present in kobotuils but I'd be nice to have accurate image support (not that I use it currently)

**Comment by NiLuJe** [comment]
@marcus-crane: Short of digging up what Nickel actually does (which is not always the right thing, see our previous discussions about that ;p), the perfect test-case is just feeding a large-enough PNG in a perfect aspect-ratio (IIRC, that's still something close to a multiple of 600x800, no matter the device?) to Nickel, and seeing what it actually generates ;o).

The full size one is pretty easy, as it's simply the actual screen size, which is easy enough to get at if you have access to the device in the first place.

The TL;DR is that calibre usually does the right thing (and if it doesn't on a device I own, I tend to try to get it fixed ;p).

**Comment by marcus-crane** [comment]
Thanks! I'll have a trawl through some of the past issues/PRs and see what tips are hidden in there 🤓

**Comment by pgaskin** [comment]
> There's a strong community there -- see for example [this list of codenames](https://www.mobileread.com/forums/showthread.php?t=325471)

Yes, that one's me btw.

> I clearly admit that I do not know how to dig into the Kobo firmware myself, not without spending hours learning: my reverse-engineering days in Assembly are from the previous century, around 1997 :)

I have a [gist](https://gist.github.com/pgaskin/613b34c23f026f7c39c50ee32f5e167e) where I documented some of it. The notes are probably a bit too sparse if you're just starting, though.

> In the worst-case scenario, say we have a the resolution wrong for some of these new models. It's still better than not supporting those models, and will allow any error to surface and make its way to the Github issues eventually.

Definitely.

> I'd be super grateful if you would merge and make a release with this.

OK, will do (I wasn't sure if this was complete earlier). I'll also make a new release of covergen with it since some people have been emailing me about that recently. Thanks for working on this PR!

**Review by pgaskin — APPROVED** [review]

**Comment by pgaskin** [comment]
It's been too long since I last updated this repo, so I'll need to rewrite the build scripts and update the tokens to do a proper release... I don't have time to do that today, but you still can pull the latest commit into another project with `go get github.com/pgaskin/koboutils/v2@master`.

**Comment by jucor** [comment]
Amazing, thank you @pgaskin for taking the time to read and merge this! It's now even already used in Kobo-UNCaGED , see https://github.com/shermp/Kobo-UNCaGED/issues/43 :) And thanks for the list of codenames and all the work, hadn't realised you were geek1011.

**Comment by pgaskin** [comment]
> hadn't realised you were geek1011

Yes, that was my old username until I changed it a few years ago.
